[FLINK-3941][TableAPI]Add support for UNION (with duplicate elimination)

yjshen opened the pull request:
This PR aims at adding `UNION` support in TableAPI and SQL by:
- Extending Table API with a new `union()` method
- Translate `UNION` to `Aggregate` on top of `UNION ALL` with the help of existing `UnionToDistinctRule` in `DATASET_OPT_RULES` 
- Fix the `expectedType` in `DataSetUnion` with result type coercion. 

Note: `Union` on streams cannot be supported right now. It should be build on windowing logical to remove duplicate rows.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetUnion.scala:
```diff
@@ -69,7 +73,7 @@ class DataSetUnion(
     }
 
-    planner.getCostFactory.makeCost(rowCnt, 0, 0)
+    planner.getCostFactory.makeCost(if (all) rowCnt else rowCnt * 0.1, 0, 0)
```
The cost for union should be higher than for union all. Also, `rowCnt` is the number of rows processed, not the result size.
How about this?
```
planner.getCostFactory.makeCost(
  rowCnt, 
  if (all) 0 else, rowCnt,
  if (all) 0 else rowCnt)
```

fhueske commented on flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/table/UnionITCase.java:
```diff
@@ -43,7 +43,7 @@ public UnionITCase(TestExecutionMode mode) {
 
 	@Test
-	public void testUnion() throws Exception {
+	public void testUnionAll() throws Exception {
```
Actually, I would completely remove the whole Java `UnionITCase.java` file. It tests exactly the same methods as the Scala tests (no expression parsing involved).

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/UnionITCase.scala:
```diff
@@ -54,7 +54,22 @@ class UnionITCase(
 
   @Test
-  def testTernaryUnion(): Unit = {
+  def testUnion(): Unit = {
```
There were some concerns about the build time of the `flink-table` module lately. In the past, we added quite a few integration tests without thinking about the added coverage vs. build time trade-off.

I would propose to replace the `testUnionWithFilter()` and `testUnionWithJoin()` tests with your new methods because these methods do not add to test coverage, IMO.

fhueske commented:
Thanks for the PR, @yjshen. Looks good except for a few minor comments.
Can you also add one test method to the Scala SQL UnionITCase, such that the SQL side is also covered?

Union on streams cannot be supported right now. It would need a lot of the windowing logic to deduplicate rows.

Thanks, Fabian

yjshen commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateMapFunction.scala:
```diff
@@ -44,7 +44,7 @@ class AggregateMapFunction[IN, OUT](
   override def map(value: IN): OUT = {
     
-    val input = value.asInstanceOf[Row]
+    val input = value.asInstanceOf[Product]
```
A `scala.Tuple` is not an instance of `Row`

fhueske replied:
Currently aggregates do only support `Row`, because the aggregate code is not generated yet.
`DataSetAggregate` enforces `Row` as input type (see `DataSetAggregate` line 99), so `value.asInstanceOf[Row]` should be safe. 

Did you observe a problem with this cast?

yjshen replied:
In `Execution mode = COLLECTION, Table config = EFFICIENT` for `testUnion`, the `value` is of `scala.Tuple3` type, not work as expected?

yjshen replied:
Ah, I think I get why this happen, will fix this.

yjshen commented:
Hi @fhueske , thanks for the review! The changes are:
- Remove the whole `java/batch/table/UnionITCase.java` 
- Remove `testUnionWithFilter`, `testUnionWithJoin` and `testUnionWithAggregate`
- Add SQL Union Test.

Looks better now?

fhueske commented:
Changes look good. Can you also please update the supported feature set in the docs (`docs/apis/table.md`)?

Should be good to merge once that is done. 
Thanks, Fabian

yjshen commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetUnion.scala:
```diff
@@ -69,16 +73,23 @@ class DataSetUnion(
+    if (all) {
+      leftDataSet.union(rightDataSet).asInstanceOf[DataSet[Any]]
+    } else {
+      leftDataSet.union(rightDataSet).distinct().asInstanceOf[DataSet[Any]]
```
In `DATASET_OPT_RULES`, `UnionToDistinctRule` substitute `Union` with `UnionAll` followed by an `Aggregate`, therefore this branch doesn't actually get executed.

fhueske replied:
Oh, yes. Completely forgot about that rule... 😊 
So, we already supported the non-all union for SQL. Only the Table API was missing the `union()` method.
I think there are two ways to continue: 
- remove the `UnionToDistinctRule` from `FlinkRuleSets`
- revert the changes on `DataSetUnion` (except of pushing down the `expectedType`) and `DataSetUnionRule`.

I am fine either ways.

yjshen replied:
I have question here on computeSelfCost, when will this method called? Is that possible we are using union's computeSelfCompute first, before UnionToDistinctRule are called?  I was trying to understand this last night but didn't come up with an idea.

fhueske replied:
This method should called by the optimizer when a new `DataSetUnion` node is created to estimate the cost of the subplan (the new node + all recursive input nodes). If there is a cheaper plan that does the same thing, the more expensive plan is discarded.

So, you have a plan with a non-union all operator. The optimizer knows the cost of this plan. Then, the `UnionToDistinctRule` is called and a new union + distinct operators are created. For both, the `computeSelfCost` method is called to compute the cost estimate of the plan and then the cheaper of both plans is preserved. (This is a bit simplified, because the optimization rules are applied on `LogicalRel` nodes but the cost estimation happens on the `DataSetRel` nodes).

yjshen commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetUnion.scala:
```diff
@@ -69,16 +73,23 @@ class DataSetUnion(
-    val rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
-    leftDataSet.union(rightDataSet).asInstanceOf[DataSet[Any]]
+    val leftDataSet = left.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
+    val rightDataSet = right.asInstanceOf[DataSetRel].translateToPlan(tableEnv, expectedType)
```
`expectedType` is passed down to `Union`'s children, enables possible conversion to `Row` enforced by `Aggregate`.

fhueske replied:
Good catch!

fhueske replied:
Actually, this fix should be extended. If `expectedType == None`, we must ensure that both inputs provide the same type, i.e., request the type of the left input on the right input (or vice versa).

fhueske replied:
With this fix, we should be able to remove all guards for efficient type usage in the Union ITCases.
There are a few like the following:
```
if (tEnv.getConfig.getEfficientTypeUsage) {
  return
}
```

yjshen replied:
Ah yes, I didn't notice that, will do :)

yjshen commented:
@fhueske would you please take another look? Thanks.

fhueske commented:
Hi, thanks for the update and the `expectedType` fix!
Can you extend it as I described in the comment and do either of the two options regarding the `UnionToDistinctRule`?

Thanks!

yjshen commented:
Hi @fhueske , I've updated the `expectedType` fix and reverted changes in `DataSetUnion` and `DataSetUnionRule` to make the implementation clear and simple, what do you think?

fhueske commented:
Perfect! Good to merge. Thanks @yjshen!

fhueske commented:
Merging
